### PR TITLE
Use cache instead of map in the plugin promwrapper

### PR DESCRIPTION
**Pull request by 0xnogo**
Context:
A memory leak was detected in CCIP nodes, traced to the `reportEndTimes` map in `promwrapper/plugin.go`. This map stores timestamps during the `Report` method but relies on `ShouldAcceptFinalizedReport` to remove entries. However, in OCR2, if Report returns false on some nodes, `ShouldAcceptFinalizedReport` may not always run, leaving stale entries that accumulate over time.

Solution:
We replaced the `reportEndTimes` map with a cache that supports expiration. As the data is only used for logging, this change avoids memory leaks without affecting performance or critical paths. Entries are now automatically cleaned up, preventing unbounded memory growth.

Testing:
This fix was tested in the `beta-testnet` environment with the following results:
With fix: ~5 days, memory usage stabilized between 217 to 242 MB.
Without fix: ~4 days, memory usage increased from 200 to 347 MB.

### CODE REVIEW

File: core/services/ocr2/plugins/promwrapper/plugin.go [hunk 1]
```diff
@@ -5,17 +5,26 @@ package promwrapper
...
 	"github.com/prometheus/client_golang/prometheus/promauto"
 	"github.com/smartcontractkit/libocr/offchainreporting2plus/types"
 )
 
+const (
+	// DefaultExpiration is the default expiration time for cache items.
+	DefaultExpiration = 30 * time.Minute
+
+	// DefaultCleanupInterval is the default interval for cache cleanup.
+	DefaultCleanupInterval = 5 * time.Minute
```
Comment (winder) [inline core/services/ocr2/plugins/promwrapper/plugin.go]: nit: make these private

File: core/services/ocr2/plugins/promwrapper/plugin.go [hunk 2]
```diff
@@ -310,12 +326,13 @@ func (p *promPlugin) ShouldAcceptFinalizedReport(ctx context.Context, timestamp
...
 
+	key := timestampToKey(timestamp)
 	// Report latency between ShouldAcceptFinalizedReport() and ShouldTransmitAcceptedReport().
 	labelValues := getLabelsValues(p, timestamp)
-	if acceptFinalizedReportEndTime, ok := p.acceptFinalizedReportEndTimes.Load(timestamp); ok {
+	if acceptFinalizedReportEndTime, ok := p.acceptFinalizedReportEndTimes.Get(key); ok {
 		latency := float64(start.Sub(acceptFinalizedReportEndTime.(time.Time)))
 		p.prometheusBackend.SetAcceptFinalizedReportToTransmitAcceptedReportLatency(labelValues, latency)
-		p.acceptFinalizedReportEndTimes.Delete(timestamp)
+		p.acceptFinalizedReportEndTimes.Delete(key)
```
Comment (mateusz-sekara) [inline core/services/ocr2/plugins/promwrapper/plugin.go]: I wonder if need to take care of deletion at all if we have eviction strategy in place

**Review by mateusz-sekara — DISMISSED** [review]

File: core/services/ocr2/plugins/promwrapper/plugin.go [hunk 3]
```diff
@@ -223,13 +232,17 @@ func New(
...
+		name:                          name,
+		chainType:                     chainType,
+		chainID:                       chainID,
+		oracleID:                      fmt.Sprintf("%d", config.OracleID),
+		configDigest:                  common.Bytes2Hex(config.ConfigDigest[:]),
+		prometheusBackend:             prometheusBackend,
+		queryEndTimes:                 cache.New(DefaultExpiration, DefaultCleanupInterval),
+		observationEndTimes:           cache.New(DefaultExpiration, DefaultCleanupInterval),
+		reportEndTimes:                cache.New(DefaultExpiration, DefaultCleanupInterval),
+		acceptFinalizedReportEndTimes: cache.New(DefaultExpiration, DefaultCleanupInterval),
```
Comment (winder) [inline core/services/ocr2/plugins/promwrapper/plugin.go]: out of curiosity, is there a reason we can't use the built-in [histogram](https://prometheus.io/docs/practices/histograms/) type to collect this sort of data with a fixed memory size?

**Review by winder — DISMISSED** [review]
Looks good. Just one question about the approach.

File: core/services/ocr2/plugins/promwrapper/plugin.go (hunk 3 above)
Comment (0xnogo) [reply core/services/ocr2/plugins/promwrapper/plugin.go]: The histogram itself can't measure time between two points. If I am getting it correctly, it's just a collection of counters that record how many observations fall into predefined duration buckets. We need the timestamp cache to actually calculate the elapsed time between one OCR step and the next one.

Comment (winder) [reply core/services/ocr2/plugins/promwrapper/plugin.go]: Ok, I thought it was measuring the steps themselves not the time between steps. That makes it more complicated.

File: core/services/ocr2/plugins/promwrapper/plugin.go (hunk 2 above)
Comment (0xnogo) [reply core/services/ocr2/plugins/promwrapper/plugin.go]: You are right, there is no strong benefit at deleting as eviction is happening every 30 minutes.

File: core/services/ocr2/plugins/promwrapper/plugin.go [hunk 4]
```diff
@@ -343,3 +352,13 @@ func (p *promPlugin) Close() error {
 
 	return p.wrapped.Close()
 }
+
+func (p *promPlugin) setEndTime(timestamp types.ReportTimestamp, cache *cache.Cache) {
+	cache.SetDefault(timestampToKey(timestamp), time.Now().UTC())
+}
+
+func timestampToKey(timestamp types.ReportTimestamp) string {
+	jsonData, _ := json.Marshal(timestamp)
```
Comment (jmank88) [inline core/services/ocr2/plugins/promwrapper/plugin.go]: Unchecked error.

The linter should have flagged this :thinking:

Comment (makramkd) [reply core/services/ocr2/plugins/promwrapper/plugin.go]: Not sure we need a JSON marshal, I think the report timestamp can be safely marshaled to a string without any errors:

```golang
key := fmt.Sprintf("%x_%d_%d", timestamp.ConfigDigest[:], timestamp.Epoch, timestamp.Round)
```

Comment (mateusz-sekara) [reply core/services/ocr2/plugins/promwrapper/plugin.go]: Agree with @makramkd, no need to run json marshaller, `sprintf` will probably be faster and sufficient for that case

**Review by mateusz-sekara — APPROVED** [review]

**Review by jmank88 — APPROVED** [review]

**Pull request merged**